### PR TITLE
Press F to kill

### LINKED ISSUE #16: [FEATURE REQ] Add keybinding to kill processes
Generally I use a system profiler to monitor processes taking up large amounts of system resources, which I would usually need to kill. Having a keybinding to kill a process from the proc page would be very helpful in this regard

**Describe the solution you'd like**
A simple single key keybinding to kill a highlighted process from within the proc page

**Describe alternatives you've considered if any**
None

### PR DESCRIPTION
**Pull request by Samyak2**
# Description

Added two keybindings to kill the selected process in `grofer proc`:
  - `K` (shift+K)
  - `<F9>` - because [htop](https://github.com/htop-dev/htop/) uses this keybind

Fixes #16 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# Checklist:

- [x] I have read the [contribution guidelines](https://github.com/pesos/grofer/blob/master/CONTRIBUTING.md) and followed it as far as possible. 
- [x] I have performed a self-review of my own code (if applicable)
- [x] I have commented my code, particularly in hard-to-understand areas (if applicable)
- [x] I have run `go fmt` on my code ([reference](https://blog.golang.org/gofmt))
- [x] My changes generate no new warnings 
- [x] Any dependent and pending changes have been merged and published

### CODE REVIEW

**Review by MadhavJivrajani — APPROVED** [review]
Looks good!
PS love the PR title xD

**Review by Gituser143 — COMMENTED** [review]
@Samyak2 is it necessary to get the `info` for the list of processes everytime data is sent through the channel? Would it be possible to maybe fetch this only when we need to kill the process? 

PS: This is just a thought, might not work though :P

**Comment by MadhavJivrajani** [comment]
That's interesting, maybe you could do something like as follows,
You could get the PID of the process as the first element of the space-separated string: 
```
# not go code
pid = myPage.BodyList.Rows[myPage.BodyList.SelectedRow].Split(...)
```

After getting the PID, you could get the process type from `gopsutil` by calling the [NewProcess](https://godoc.org/github.com/shirou/gopsutil/process#NewProcess) function:
```go
procToKill, err := NewProcess(pid)
```
And on this `process` type you could call the `Kill` function as you already have.

We will have to implement the error handling workflow as implemented in #53 , this can come under a separate PR.

**Review by metonymic-smokey — APPROVED** [review]

**Comment by Gituser143** [comment]
@MadhavJivrajani That sounds good, @Samyak2 what do you think?

**Comment by Samyak2** [comment]
Sounds good. I thought we would need the `procs` list for some other operations, but since we'll need details of only one process Madhav's suggestion is a better idea. I'll make the changes.

**Comment by Samyak2** [comment]
I have added some error handling, but I'm not sure if that's the right way to do it.

**Review by MadhavJivrajani — COMMENTED** [review]
The changes to execute the `Kill()` function look good. However, the error handling part might work in the sense that it will terminate the execution of the program, but there the error won't be propagated up the function call lineage in order for it to be visible/gracefully terminate. If you're interested in working on the error handling part of this as well then we'd be more than happy to help you through it!

**Comment by Samyak2** [comment]
Yes! I would like to work on it. I'm looking at #53 to get some idea.

I'll also fix the formatting issue.

**Comment by MadhavJivrajani** [comment]
Awesome!
Let us know if you need any help :smile:

**Comment by Samyak2** [comment]
I noticed that there's no error handling in `cmd/proc.go`, so I'll have to add it similar to #53 with the following changes:
 - Make an `errgroup`, call the goroutines through the error group and then collect the error and print.
 - Change the `process` and `procGraph` functions to  use `context.Content` instead of `endChannel` and to return `error`.

Is this all there is to it or am I missing something?

**Comment by MadhavJivrajani** [comment]
That sounds about right. 
There is [another place](https://github.com/pesos/grofer/blob/c10ba49a1b8d5dd01418ef77f0b68755971690c6/src/display/process/procGraphs.go#L57) (unrelated to this PR) where error handling needs to be implemented, you can take it up along with this PR itself or not, either way, your approach LGTM.

**Comment by Samyak2** [comment]
I have implemented error handling as in #53, but I've run into an issue. When an error is encountered (or when q is pressed), grofer exits and shows the error message but the UI does not seem to exit gracefully - cursor is invisible in the prompt and the UI remains in the terminal output.

image

**Comment by MadhavJivrajani** [comment]
Hi,
I was able to figure out the issue with that, it seems to be an error from our end. For some reason, for the function `AllProcVisuals()` a `defer ui.Close()` seemed to be missing because of which the graceful exit wasn't happening. The reason it was working so far was that we were issuing an `os.Kill` whenever the events `q` or `<C c>` took place. Placing a `defer ui.Close()` after line 133 in your current code should do the trick! My apologies for the confusion.

**Comment by Samyak2** [comment]
That works! Thanks :smile:

**Comment by Samyak2** [comment]
I have a concern wrt this PR. The shortcut "F" is a little too close to "G" (for scrolling to the bottom), it might lead to accidentally killing a process (I think this happened to me twice, had to restart/relogin because the window manager stopped working).

I was thinking of changing it to "K" (or any other alternative that might better). What do you think?

(RIP title though)

**Comment by MadhavJivrajani** [comment]
Ah damn 
That makes sense, yeah that sounds good go ahead with `K` itself.

**Comment by Samyak2** [comment]
I ran into another issue :sweat:. Seems to be some kind of race condition.
To reproduce:
 - Start grofer proc with a high refresh rate. `./grofer proc -r 5000`
 - Start a non-terminating process in another terminal: Ex: `cat`
 - Go to the newly created `cat` process in grofer
 - Terminate (ctrl-c) the `cat` process
 - Wait a second or two and try to kill the cat process ('F')
 - See this error (repeat if it says "process does not exist")

image

It seems that the `SelectedRow` is not updated when the list is updated. This can sometimes lead to the wrong process being killed. For example, repeat the above steps but run `cat` before running grofer and it will lead to `grofer` getting killed instead of cat (assuming no other processes were created during this time)

image

**Comment by Gituser143** [comment]
@Samyak2 `process.Kill()` returns an error type, maybe we could use this in anyway? Apart from this, a simple call to `updateUI()` at the beginning of the case statement to kill might just help.

**Comment by Gituser143** [comment]
@Samyak2 And as for the second issue when `cat` is killed externally first and then attempted to be killed through `grofer`, a naive approach which I feel might work is something like this:

```go
case "F", "<F9>":
  initialRow := myPage.BodyList.Rows[myPage.BodyList.SelectedRow]
  // get initial PID from the data
  initialPid64, err := strconv.ParseInt(strings.SplitN(initialRow, " ", 2)[0], 10, 32)
  // handle err
  initialPid := int32(initialPid64)
  updateUI()
  // get row selected by grofer
  row := myPage.BodyList.Rows[myPage.BodyList.SelectedRow]
  // get PID from the data
  pid64, err := strconv.ParseInt(strings.SplitN(row, " ", 2)[0], 10, 32)
  pid := int32(pid64)
  // get process and kill it
  procToKill, err := proc.NewProcess(pid)
  if err == nil && pid == initialPid {
    procToKill.Kill()
  }
```

**Comment by Samyak2** [comment]
> @Samyak2 `process.Kill()` returns an error type, maybe we could use this in some way?

Oh yes, I must have missed that while doing the error handling.

@Gituser143 I tried your solution, it did not fix either of the issues - index of out range and the issue of killing the wrong process.

```go
case "K", "<F9>":
	initialRow := myPage.BodyList.Rows[myPage.BodyList.SelectedRow]
	// get initial PID from the data
	initialPid64, err := strconv.ParseInt(strings.SplitN(initialRow, " ", 2)[0], 10, 32)
	if err != nil {
		return fmt.Errorf("Failed to get PID of process: %v", err)
	}
	initialPid := int32(initialPid64)

	updateUI()

	row := myPage.BodyList.Rows[myPage.BodyList.SelectedRow]
	// get PID from the data
	pid64, err := strconv.ParseInt(strings.SplitN(row, " ", 2)[0], 10, 32)
	if err != nil {
		return fmt.Errorf("Failed to get PID of process: %v", err)
	}
	pid := int32(pid64)

	// get process and kill it
	procToKill, err := proc.NewProcess(pid)
	if err != nil {
		return fmt.Errorf("Failed to get process with PID %d: %v", pid, err)
	} else if pid == initialPid {
		err = procToKill.Kill()
		if err != nil {
			return fmt.Errorf("Failed to kill process with PID %d: %v", pid, err)
		}
	}
```

**Comment by Gituser143** [comment]
@Samyak2 I tried this out in a similar but sloppy way. Here you're checking if `err != nil` and if so returning/printing error. Instead, avoid that case entirely and only work on cases if `err == nil`. 

```go
case "F", "<F9>":
				initialRow := myPage.BodyList.Rows[myPage.BodyList.SelectedRow]
				// get PID from the data
				initialPid64, err := strconv.ParseInt(strings.SplitN(initialRow, " ", 2)[0], 10, 32)
				if err != nil {
					return fmt.Errorf("Failed to get PID of process: %v", err)
				}
				initialPid := int32(initialPid64)
				updateUI()
				row := myPage.BodyList.Rows[myPage.BodyList.SelectedRow]
				// get PID from the data
				pid64, err := strconv.ParseInt(strings.SplitN(row, " ", 2)[0], 10, 32)
				if err != nil {
					return fmt.Errorf("Failed to get PID of process: %v", err)
				}
				pid := int32(pid64)
				// get process and kill it
				procToKill, err := proc.NewProcess(pid)
				if err == nil && pid == initialPid {
					procToKill.Kill()
				}
			} 
```

**Comment by Samyak2** [comment]
> Here you're checking if `err != nil` and if so returning/printing error. Instead, avoid that case entirely and only work on cases if `err == nil`.

@Gituser143 Ah right, I missed that.

That does help when the process has already exited, but if the row list is updated without updating the `SelectedRow` it still fails with a "index out of range" panic (even with `UpdateUI`)

**Comment by Gituser143** [comment]
@Samyak2 the confirmation sounds like a good idea, I hadn't brought that up previously because I'm not sure if we can do that without creating another widget entirely. If you feel this can be done without overly complicating it, I'm all for it.

PS: could you tell me how to replicate the index out of range error? I can't seem to be able to do so.

**Comment by Samyak2** [comment]
I'm not familiar with the UI library, so I'm not sure if I can do it quickly. (Perhaps we could open a new issue for this?)

To reproduce the index out of range error (from the commit I just pushed):
 - `./grofer proc -r 5000`
 - Run `cat` on another terminal
 - Find the `cat` process in grofer
 - Ctrl-C the `cat` process on the other terminal
 - In grofer: Spam the kill key ('K') on the `cat` process until it exits with a panic.

**Comment by Gituser143** [comment]
@Samyak2 I just looked at the implementation of kill in gotop. Here, exec is used to kill using the kill command itself. Might or might not be helpful :P

**Comment by Gituser143** [comment]
> I'm not familiar with the UI library, so I'm not sure if I can do it quickly. (Perhaps we could open a new issue for this?)
> 
> To reproduce the index out of range error (from the commit I just pushed):
> 
>     * `./grofer proc -r 5000`
> 
>     * Run `cat` on another terminal
> 
>     * Find the `cat` process in grofer
> 
>     * Ctrl-C the `cat` process on the other terminal
> 
>     * In grofer: Spam the kill key ('K') on the `cat` process until it exits with a panic.

Also, if we update the UI right after the kill would it help in anyway? Although naive, I feel that just might help.

**Comment by Gituser143** [comment]
@Samyak2 I just realised, maybe we could use a waitGroup to ensure nothing else happens while killing. This way even if keys are spammed after the first F9, it won't even be recorded. What do you think?

PS: I think 'k' for kill isn't appropriate as it's being used currently to scroll. I only realised this, when I couldn't kill procs now

**Comment by Samyak2** [comment]
> @Samyak2 I just realised, maybe we could use a waitGroup to ensure nothing else happens while killing. This way even if keys are spammed after the first F9, it won't even be recorded. What do you think?

I'm not entirely sure how a wait group works yet, but is your suggestion is to do `wg.Wait` before killing?

> PS: I think 'k' for kill isn't appropriate as it's being used currently to scroll. I only realised this, when I couldn't kill procs now 

The key for killing is `K` (shift K) and not `k` 😄

**Comment by Gituser143** [comment]
> The key for killing is `K` (shift K) and not `k` smile
I completely missed that, my bad. 

> I'm not entirely sure how a wait group works yet, but is your suggestion is to do `wg.Wait` before killing?

We wait while it's killing. A waitGroup is essentially a semaphore. We set it's value to 1 (`wg.Add(1)`), call a routine to kill the process and `wg.Wait()` till that routine is finished (call `wg.Done()` inside the routine). This might help avoiding key spam, but it certainly won't fix the index out of range issue (when we try to get the selected row, but the row was just removed due to the process exiting).

**Comment by MadhavJivrajani** [comment]
Calling `updateUI()` won't solve the issue here since it's only resizing the widgets. IMO you will have to call the [`Processes()`](https://godoc.org/github.com/shirou/gopsutil/process#Processes) function as done [here](https://github.com/pesos/grofer/blob/c10ba49a1b8d5dd01418ef77f0b68755971690c6/src/process/serveData.go#L51) and then update the rows as done in line 220 of your current code. 

A cleaner way to check if a process exists when you're calling the `Kill()` function would be to use the [`PidExists()`](https://godoc.org/github.com/shirou/gopsutil/process#PidExists) function and handle the error in a similar way as done so far. Only if the PID exists then kill the process, if not then don't do anything, and only after doing this, call the `Processes()` function and update the list of processes so that it doesn't have to wait till the next tick.

**Comment by Samyak2** [comment]
>  IMO you will have to call the [`Processes()`](https://godoc.org/github.com/shirou/gopsutil/process#Processes) function as done [here](https://github.com/pesos/grofer/blob/c10ba49a1b8d5dd01418ef77f0b68755971690c6/src/process/serveData.go#L51) and then update the rows as done in line 220 of your current code.

The UI updating as soon as you try to kill a process does not seem like the best solution, but I guess it's better than waiting for the next UI update and accidentally killing the wrong process.

> A cleaner way to check if a process exists when you're calling the `Kill()` function would be to use the [`PidExists()`](https://godoc.org/github.com/shirou/gopsutil/process#PidExists) function and handle the error in a similar way as done so far. 

This is a better solution than just ignoring the error from `Kill()` :smile: 

> Only if the PID exists then kill the process, if not then don't do anything, and only after doing this, call the `Processes()` function and update the list of processes so that it doesn't have to wait till the next tick.

With this, the issue of killing the wrong process still exists. I think it would be better to update the UI after K is pressed before doing anything else.

(Sorry for the late response :disappointed:)

**Comment by MadhavJivrajani** [comment]
> With this, the issue of killing the wrong process still exists. I think it would be better to update the UI after K is pressed before doing anything else.
Yes, you're right, my bad.

**Comment by Gituser143** [comment]
@Samyak2 An approach we could take would be to get the row as soon as F is pressed, we can then extract the PID from that row. Once this happens, we can call the `pause()` function to stop updating UI. 

Once we get the PID from that row, we can confirm with the user if the process they want to kill matches the one we caught. Followed by a (y/n) prompt through a text box widget. 

On confirmation, we can check if the process still exists and kill it.

What are your thoughts on this?

**Comment by MadhavJivrajani** [comment]
Seems forced tbf, it shouldn't be more work than it has to be

**Comment by Samyak2** [comment]
@Gituser143 I like the idea. A confirmation is, IMO, quite important for this feature since it's `SIGKILL`ing a process.

`htop` does it in a similar way except that the selection is moved as the list is getting updated.

**Comment by MadhavJivrajani** [comment]
Confirmation of some form is important, agreed, but if the same issue as before occurs, that is if the wrong process is selected for killing, then prompting that and the user selecting `no` and then repeating it, that might not be what should be done to guarantee the best experience, and if the same thing happens more than twice it could also count as borderline annoying for the user.

I might be misunderstanding the whole thing, please correct me if I've gotten it wrong.

**Comment by Samyak2** [comment]
That is true, it will get annoying if that happens too many times in a row, but I don't think there's a simple solution that can fix this. For context, `htop` (i'm sorry that's the only other profiler i know :laughing:) behaves similarly - the process list changes quickly and a wrong process might get selected but since there is a confirmation (selection of signal to send), the wrong process is never killed.

For now, I was thinking of doing the following:
 - Update the process list when `K` is pressed
 - If the PID exists, display a confirmation
 - If confirmed, kill if the PID still exists

(Also, another thing I noticed in `htop` was that, when a process is selected and it dies while it was selected, the selection is automatically cleared)

**Comment by MadhavJivrajani** [comment]
I see what you mean, `htop` also provides a list of signals that could help kill the process. That's what should be done as well instead of directly send a `SIGKILL`. Afaik `gopsutil` does not provide functionality for other signals other than `SIGKILL`, but implementing that is straightforward. If we can decide on this approach of killing a process (i.e. providing signal options to the users), then I can open a separate issue for implementing kill signal functionality and this PR can be labeled as long term.

**Comment by Gituser143** [comment]
> Confirmation of some form is important, agreed, but if the same issue as before occurs, that is if the wrong process is selected for killing, then prompting that and the user selecting no and then repeating it, that might not be what should be done to guarantee the best experience, and if the same thing happens more than twice it could also count as borderline annoying for the user.

This won't be an issue. Because as soon as we get the row, we can kill slowly and if it doesn't exist, return gracefully. Since we update UI right after this above return/process killing. The user won't be clicking wrong PIDs again and again

**Comment by Samyak2** [comment]
> I see what you mean, `htop` also provides a list of signals that could help kill the process. That's what should be done as well instead of directly send a `SIGKILL`. Afaik `gopsutil` does not provide functionality for other signals other than `SIGKILL`, but implementing that is straightforward.

According to the docs, `gopsutil` only supports 4 types of signals being sent to a process ([source](https://godoc.org/github.com/shirou/gopsutil/process#Process.SendSignal)). Is there another way of sending a signal to a process?

> If we can decide on this approach of killing a process (i.e. providing signal options to the users), then I can open a separate issue for implementing kill signal functionality and this PR can be labeled as long-term.

How about this: for this PR, I'll implement a simple confirmation - press K once to select a process, pause UI, press K again to kill (if it still exists).
You can then open an issue to extend this to any signal instead of just `SIGKILL` where we'll need a separate UI component with the list of signals.

**Comment by Samyak2** [comment]
@Gituser143 @MadhavJivrajani what do you think of the changes? Is it a good way to do this?

(also, the `hacktoberfest-accepted` label isn't required, this PR was opened before October)

File: src/display/process/allProcs.go
Comment (Gituser143) [inline src/display/process/allProcs.go:161]: @Samyak2 In this function since you're using the `sync.Once` variable, will it really call `updateUI`? Because the once variable is used to render the very first data coming in from the data channel.

Comment (Gituser143) [reply src/display/process/allProcs.go:161]: Apart from this, the `updateProcs()` function is called at the very beginning of the kill section when `K` is pressed. I'm not sure what your reasoning is to call that there. Please let me know?

Comment (Samyak2) [reply src/display/process/allProcs.go:161]: @Gituser143 you are right, it won't actually call `updateUI`. I had copied this part from `case data := <-dataChannel:` and forgot to remove the Once line 😅.
The reasoning behind calling `updateProcs` at the beginning of the kill section was to ensure that the process selected for kill actually exists. But upon further testing, I found that this leads to a different process getting selected for killing.

IMO a better approach would be doing this:
 - After a process is selected for killing, monitor it in every update
 - if the PID still exists, do nothing
 - if the PID does not exist, remove selection

File: src/display/process/allProcs.go
Comment (Gituser143) [inline src/display/process/allProcs.go:288]: @Samyak2 This method of checking is accurate but the issue is we'll never get to this section of code as the above statements are in an infinite loop. The only time we exit from that loop is when we exit the program entirely! What you're doing in the else clause of `if !killSelected` suffices I feel.

File: src/display/process/allProcs.go
Comment (Gituser143) [inline src/display/process/allProcs.go]: @Samyak2 Wouldn't returning this error hinder with the UI? Because in the `proc.Kill()` function, it checks again if the PID exists and then sends a signal. I'm against returning an error here if `proc.kill()` fails as it would unnecessarily stop the UI and quit. I'd drop the check entirely.

Maybe let's wait for @MadhavJivrajani for a better opinion?

**Review by Gituser143 — CHANGES_REQUESTED** [review]

File: src/display/process/allProcs.go
Comment (Samyak2) [reply src/display/process/allProcs.go:288]: This is inside the loop xD, right after the `select`

File: src/display/process/allProcs.go
Comment (Samyak2) [reply src/display/process/allProcs.go]: I'm returning errors in two places (inside kill):
 - If parsing the PID fails: this shouldn't happen in any condition, unless there's a non-numeric PID printed somehow. I think we should keep this since this error will mean that something is seriously wrong.
 - Process kill fails: this could mean permission issues. We could probably ignore this, but there must be some way to let the user know that it failed.

File: src/display/process/allProcs.go
Comment (Gituser143) [reply src/display/process/allProcs.go:288]: Oops, my bad!

File: src/display/process/allProcs.go
Comment (Gituser143) [reply src/display/process/allProcs.go]: @Samyak2 

> but there must be some way to let the user know that it failed.

I'd prefer this than closing the UI. Maybe render a temporary textbox to deliver the message?

Comment (MadhavJivrajani) [reply src/display/process/allProcs.go]: Yes, I'd also prefer some way of letting the user know and not closing the UI. Maybe instead of the textbox, we could have something simpler like making the color of the selected row as red or something of the sort and we could mention the meaning of doing this as part of #42

Comment (Gituser143) [reply src/display/process/allProcs.go]: I thought of the same thing. But I thought it'll hinder experience. Considering the user will have to keep popping the help text open just to understand what's going on. Maybe we could get the row to turn red for now and maybe open another issue to address the delivery of the message?

Comment (Samyak2) [reply src/display/process/allProcs.go]: Yes, sounds good. Currently, the row is highlighted in red when it is selected for killing, I'll change the colors.

**Comment by Gituser143** [comment]
@Samyak2 On pressing `K` for the first time, if we wait for more than `refreshRate` amount of time, the selection is unhilighted. Could you please look into this?

**Comment by Samyak2** [comment]
Oops

**Review by Gituser143 — APPROVED** [review]

**Comment by Gituser143** [comment]
@Samyak2 This looks good! There are a few conflicts that have to be resolved though. Could you please look into that?
Conflicting files are:
* cmd/proc.go
* src/display/process/allProcs.go
* src/display/process/procGraphs.go
* src/process/serveData.go

**Comment by Samyak2** [comment]
Yes, I'll rebase on master

**Comment by Samyak2** [comment]
^ Rebased on upstream/master and squashed a few commits

**Review by Gituser143 — APPROVED** [review]

**Comment by Gituser143** [comment]
@Samyak2 Everything looks good! Great work! I'll wait for @MadhavJivrajani or @metonymic-smokey to merge if they have no issues.

**Comment by Gituser143** [comment]
@Samyak2 There seems to be 2 issues with `grofer proc -p PID`.

1. It's missing a `ui.Render(myPage.grid)` at the end of the cases. This causes a significant lag while scrolling through the child process list. Refer to #60 to see what's missed.

2. When trying to exit from the UI, it doesn't close cleanly. There's a signifiant delay there too.

**Comment by Gituser143** [comment]
@Samyak2 there seems to be some issue with `grofer proc -p PID`. The UI doesn't exit properly. Could you please look into this?

**Comment by Samyak2** [comment]
Right, the UI was not being closed in `ProcVisuals`.

Also, I noticed that it takes a second or two to exit. Do you know why?

**Comment by Gituser143** [comment]
> Also, I noticed that it takes a second or two to exit. Do you know why?

I was looking at it. I wasn't able to figure that out. I'll keep at it. I'll update you if I figure something out.

**Review by Gituser143 — APPROVED** [review]

**Comment by Gituser143** [comment]
@MadhavJivrajani @metonymic-smokey Plis review once if there's anything missing and merge.

@Samyak2 this is great work! Thanks for the PR!.
PS: You could raise a seperate issue for the UI closing delay.

**Review by MadhavJivrajani — APPROVED** [review]
@Samyak2 thanks for your contribution!
It was definitely a very interesting conversation that evolved during the course of the PR :smile: 
@Gituser143 LGTM, merge at your leisure :+1:

**Comment by Samyak2** [comment]
> It was definitely a very interesting conversation that evolved during the course of the PR

:100: this PR was a journey

**Pull request merged**
